File: vscode-extension/src/service/LLMService.ts
/**
 * LLM 服务
 */

import { ConfigService } from '../config/index.js';
import { Context } from './ContextService.js';

/**
 * LLM 错误
 */
export class LLMError extends Error {
    constructor(
        message: string,
        public code: string,
        public statusCode?: number
    ) {
        super(message);
        this.name = 'LLMError';
    }
}

/**
 * LLM 服务
 */
export class LLMService {
    
    constructor(private config: ConfigService) {}
    
    /**
     * 发送聊天消息
     */
    async chat(message: string, context: Context): Promise<string> {
        if (!this.config.isConfigured()) {
            throw new LLMError('请先配置 API Key', 'NOT_CONFIGURED');
        }
        
        const messages = this.buildMessages(message, context);
        
        const response = await fetch(`${this.config.getApiEndpoint()}/v1/chat/completions`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${this.config.getApiKey()}`,
            },
            body: JSON.stringify({
                model: this.config.getModel(),
                messages,
                max_tokens: this.config.getMaxTokens(),
            }),
        });
        
        if (!response.ok) {
            const errorText = await response.text();
            throw new LLMError(
                `API 请求失败: ${errorText}`,
                'API_ERROR',
                response.status
            );
        }
        
        const data = await response.json();
        return this.parseResponse(data);
    }
    
    /**
     * 流式聊天
     */
    async *chatStream(message: string, context: Context): AsyncGenerator<string> {
        if (!this.config.isConfigured()) {
            throw new LLMError('请先配置 API Key', 'NOT_CONFIGURED');
        }
        
        const messages = this.buildMessages(message, context);
        
        const response = await fetch(`${this.config.getApiEndpoint()}/v1/chat/completions`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${this.config.getApiKey()}`,
            },
            body: JSON.stringify({
                model: this.config.getModel(),
                messages,
                max_tokens: this.config.getMaxTokens(),
                stream: true,
            }),
        });
        
        if (!response.ok) {
            throw new LLMError('API 请求失败', 'API_ERROR', response.status);
        }
        
        if (!response.body) {
            throw new LLMError('无效的响应', 'INVALID_RESPONSE');
        }
        
        const reader = response.body.getReader();
        const decoder = new TextDecoder();
        let buffer = '';
        
        try {
            while (true) {
                const { done, value } = await reader.read();
                if (done) break;
                
                buffer += decoder.decode(value, { stream: true });
                const lines = buffer.split('\n');
                buffer = lines.pop() ?? '';
                
                for (const line of lines) {
                    const chunk = this.parseStreamChunk(line);
                    if (chunk) {
                        yield chunk;
                    }
                }
            }
        } finally {
            reader.releaseLock();
        }
    }
    
    /**
     * 构建消息
     */
    private buildMessages(message: string, context: Context): Array<{ role: string; content: string }> {
        const messages: Array<{ role: string; content: string }> = [];
        
        if (context.files.length > 0) {
            const contextText = context.files
                .map(f => `文件: ${f.path}\n\`\`\`${f.language}\n${f.content}\n\`\`\``)
                .join('\n\n');
            
            messages.push({
                role: 'system',
                content: `相关代码上下文:\n${contextText}`,
            });
        }
        
        messages.push({
            role: 'user',
            content: message,
        });
        
        return messages;
    }
    
    /**
     * 解析响应
     */
    private parseResponse(data: { choices?: Array<{ message?: { content?: string } }> }): string {
        return data.choices?.[0]?.message?.content ?? '';
    }
    
    /**
     * 解析流式块
     */
    private parseStreamChunk(line: string): string | null {
        const trimmed = line.trim();
        if (!trimmed || !trimmed.startsWith('data: ')) {
            return null;
        }
        
        const data = trimmed.slice(6);
        if (data === '[DONE]') {
            return null;
        }
        
        try {
            const chunk = JSON.parse(data);
            const content = chunk.choices?.[0]?.delta?.content;
            return content ?? null;
        } catch {
            return null;
        }
    }
}

File: idea-plugin/bin/main/com/vex/coding/service/LLMService.kt
package com.vex.coding.service

import com.vex.coding.config.ConfigService
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration

/**
 * LLM 错误类型
 */
class LLMException(
    message: String,
    val code: String,
    val statusCode: Int? = null
) : RuntimeException(message)

/**
 * LLM 服务
 */
class LLMService {
    
    private val config = ConfigService.getInstance()
    private val httpClient = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(60))
        .build()
    
    /**
     * 发送聊天消息
     */
    suspend fun chat(message: String, context: Context): String = withContext(Dispatchers.IO) {
        if (!config.isConfigured()) {
            throw LLMException("请先配置 API Key", "NOT_CONFIGURED")
        }
        
        val messages = buildMessages(message, context)
        
        val request = HttpRequest.newBuilder()
            .uri(URI.create("${config.getApiEndpoint()}/v1/chat/completions"))
            .header("Content-Type", "application/json")
            .header("Authorization", "Bearer ${config.getApiKey()}")
            .POST(HttpRequest.BodyPublishers.ofString(messages))
            .timeout(Duration.ofSeconds(config.getMaxTokens().toLong()))
            .build()
        
        try {
            val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
            
            if (!response.statusCode().toString().startsWith("2")) {
                throw LLMException(
                    "API 请求失败: ${response.body()}",
                    "API_ERROR",
                    response.statusCode()
                )
            }
            
            parseResponse(response.body())
        } catch (e: LLMException) {
            throw e
        } catch (e: Exception) {
            throw LLMException("网络错误: ${e.message}", "NETWORK_ERROR")
        }
    }
    
    /**
     * 构建请求消息
     */
    private fun buildMessages(message: String, context: Context): String {
        val messages = mutableListOf<Map<String, String>>()
        
        // 添加上下文
        if (context.files.isNotEmpty()) {
            val contextText = context.files.joinToString("\n\n") { file ->
                "文件: ${file.path}\n```${file.language}\n${file.content}\n```"
            }
            messages.add(mapOf(
                "role" to "system",
                "content" to "相关代码上下文:\n$contextText"
            ))
        }
        
        // 添加用户消息
        messages.add(mapOf(
            "role" to "user",
            "content" to message
        ))
        
        return buildString {
            append("{")
            append("\"model\": \"${config.getModel()}\",")
            append("\"messages\": ${toJson(messages)},")
            append("\"max_tokens\": ${config.getMaxTokens()}")
            append("}")
        }
    }
    
    /**
     * 解析响应
     */
    private fun parseResponse(body: String): String {
        val contentStart = body.indexOf("\"content\":\"") + 11
        val contentEnd = body.indexOf("\"", contentStart)
        
        if (contentStart > 10 && contentEnd > contentStart) {
            return body.substring(contentStart, contentEnd)
                .replace("\\n", "\n")
                .replace("\\\"", "\"")
        }
        
        return body
    }
    
    private fun toJson(messages: List<Map<String, String>>): String {
        return messages.joinToString(",", "[", "]") { msg ->
            "{\"role\":\"${msg["role"]}\",\"content\":\"${msg["content"]?.replace("\"", "\\\"")}\"}"
        }
    }
}